django_ontruck/denormalisation.py

```diff
@@ -0,0 +1,350 @@
+        self.field = field
+        self.prefetches = prefetches
+
+        super().__init__(fget=fget)
+
+
+class Denormalised:
+    """
+    A descriptor class that is used to control denormalised data.
+
+    :param model_base: The Django model_class base class that the denormalisation
+    model_class will inherit from.
+
+    :param signals: An iterable containing signals which should trigger a
+    re-denormalisation of the model_class.
+
+    ## Example
+
+    class Foo(BaseModel):
+        bar = IntegerField(default=0)
+        baz = IntegerField(default=3)
+
+        denormalised = Denormalised(
+            BaseModel, signals=(
+                post_save,
+                DecoratedSignal(
+                    SomeEvent.signal,
+                    fields=('my_expensive_prop',)
+                )
+            )
+        )
+
+        @denormalised.property(IntegerField())
+        def my_expensive_prop(self):
+            # Some expensive calculation here
+
+    foo = Foo.objects.last()
+
+    # re-denormalise the denormalised_properties of foo
+    Foo.denormalised.handler(None, instance=foo)
+
+    # calculate value of `my_expensive_prop`
+    foo.my_expensive_prop
+
+    # Or use the denormalised value
+    foo.denormalised.my_expensive_prop
+    """
+    name = InitialisedOnSetName()
+    model_class = InitialisedOnSetName()
+    parent_model_class = InitialisedOnSetName()
+
+    def __init__(
+        self,
+        model_base=Model,
+        signals=(post_save,)
+    ):
+        self.model_base = model_base
+        self.denormalised_properties = []
+        self.signals = signals
+
+    @property
+    def related_name(self):
+        return '_denormalised_{}'.format(self.name)
+
+    # Equivalent to __set_name__ in Python 3.6+
+    def contribute_to_class(self, owner, name):
+        self.name = name
+        self.parent_model_class = owner
+
+        factory = DenormalisedModelClassFactory(
+            model_base=self.model_base,
+            parent_model_class=owner,
+            name=name,
+            denormalised_properties=self.denormalised_properties,
+            parent_related_name=self.related_name,
+        )
+
+        self.model_class = factory.export()
+
+        self.connect()
+
+        setattr(owner, name, self)
+
+    def __get__(self, instance, owner):
+        if not instance:
+            return self
+
+        return getattr(instance, self.related_name)
+
+    def property(self, field, prefetches=tuple()):
+        def decorator(func):
+            denormalised_property = DenormalisedProperty(
+                func,
+                field=field,
+                prefetches=prefetches
+            )
+
+            self.denormalised_properties.append(denormalised_property)
+
+            return denormalised_property
+
+        return decorator
+
+    def handler(self, fields=None):
+        return Denormaliser(self.model_class, fields=fields)
+
+    def connect(self):
+        for signal in self.signals:
```

Comment:
Should we speak about events that is our django-ontruck idiom for our extended django signal?

Reply:
This was kind of my thinking with `DecoratedSignal`: keep the api related to `django` as opposed to `django-ontruck` but provide tools to make it compatible with events.

I think the difficulty is that our events don't have a standardised interface and so will almost always need some kind of adapter in order to be compatible.

Reply:
Ok, so I've renamed the `DecoratedSignal` to `SignalConnector` and added a class build on top of it called `EventConnector` that is designed to work with the `django-ontruck` `Event` api.